elf_reader.go

```diff
@@ -1067,6 +1094,96 @@ func (ec *elfCode) loadDataSections(maps map[string]*MapSpec) error {
 
 		maps[sec.Name] = mapSpec
 	}
+
+	if ec.btf != nil {
+		var ds *btf.Datasec
+		if err := ec.btf.TypeByName(".kconfig", &ds); err != nil {
+			if errors.Is(err, btf.ErrNotFound) {
+				return nil
+			}
+			return err
+		}
+
+		maps[".kconfig"] = &MapSpec{
+			Name:       ".kconfig",
+			Type:       Array,
+			KeySize:    uint32(4),
+			MaxEntries: 1,
+			Flags:      unix.BPF_F_RDONLY_PROG | unix.BPF_F_MMAPABLE,
+			Key:        &btf.Void{},
+			Value:      ds,
+			ValueSize:  ds.Size,
+		}
+	}
+
+	return nil
+}
+
+func (ec *elfCode) populateDataSections() error {
+	if ec.maps == nil {
+		return fmt.Errorf("elfCode map of MapSpec should not be nil")
+	}
+
+	return ec.populateKconfigSection()
+}
+
+func (ec *elfCode) populateKconfigSection() error {
+	mapSpec, ok := (*ec.maps)[".kconfig"]
+	if !ok {
+		// If .kconfig map does not exist it just means the code does not have any
+		// variable inside .kconfig data section.
+		return nil
+	}
+
+	ds, ok := mapSpec.Value.(*btf.Datasec)
+	if !ok {
+		return fmt.Errorf(".kconfig value must be of type btf.Datasec")
+	}
+
+	data := make([]byte, 0, ds.Size)
+	for _, vsi := range ds.Vars {
+		v, ok := vsi.Type.(*btf.Var)
+		if !ok {
+			return fmt.Errorf("info should point to a variable")
+		}
+
+		s, err := btf.Alignof(v.Type)
+		if err != nil {
+			return fmt.Errorf("getting size of kconfig variable %s: %w", v.TypeName(), err)
+		}
+		d := make([]byte, s)
+
+		switch n := v.TypeName(); n {
+		case "LINUX_KERNEL_VERSION":
+			if s < 4 {
+				return fmt.Errorf("kconfig variable %s must be at least u32, got %d", n, s)
+			}
+			kv, err := internal.KernelVersion()
+			if err != nil {
+				return fmt.Errorf("getting kernel version: %w", err)
+			}
+			internal.NativeEndian.PutUint32(d, kv.Kernel())
+
+		default:
+			return fmt.Errorf("unsupported kconfig variable: %s", n)
+		}
+
+		// We need to change Linkage for extern to global to avoid problems while
+		// marshaling the map value.
+		v.Linkage = btf.GlobalVar
+
+		data = append(data, d...)
+	}
+
+	length := uint32(len(data))
+
+	// We need to update DataSec size to avoid problems while marshaling the map
+	// value.
+	ds.Size = uint32(len(data))
+
+	mapSpec.ValueSize = length
```

Comment:
Let's not initialize ValueSize in `loadDataSections` if we need to fix it up here anyway.